### ISSUE TITLE
[Bug]: GLM-5.3-Flash degenerates into a constant-token loop with PP=2 + MTP + prefix caching on sm_86 (upstream fix exists: vllm#55506)

### ISSUE THREAD

**Issue by noah80** [wtdcode/vllm-backport#96]
<html><body>
<html><head></head><body><h1>[Bug]: GLM-5.3-Flash degenerates into a constant-token loop with PP=2 + MTP + prefix caching on sm_86 (upstream fix exists: vllm#55506)</h1>
<h2>Summary</h2>
<p>Serving <code>GLM-5.3-Flash</code> (FP8, <code>Glm5NextForConditionalGeneration</code>) across two
nodes at TP=8 / PP=2 with MTP-3 and prefix caching, requests intermittently
degenerate into an unbounded constant-token loop. Output is prompt-independent
once it starts and never recovers. The request returns HTTP 200; nothing is
logged.</p>
<p>This appears to be the same fault as
<a href="https://github.com/vllm-project/vllm/pull/55506">vllm-project/vllm#55506</a>
("Index mamba spec-decode block tables by request slot, not batch row"), which
was reproduced on Qwen3.8-Flash-Next. That PR notes the failure is loud on
sm_121 as an illegal memory access but <strong>silent on sm_80 where the misdirected
read stays mapped</strong>. These are A40s (sm_86) and the behaviour is the silent
variant: corrupted output, no crash.</p>
<p>Filing here because the fork carries the affected hybrid/spec-decode paths and
has already taken related fixes (#63, #77, #49), and because GLM-5.3-Flash may
need the same treatment the Qwen path got.</p>
<h2>Environment</h2>
<pre><code>image      lazymio/vllm-backport:latest-sm86  (+ ray[default] layered on top;
                                               upstream image ships no ray)
fingerprint vllm-0.13.1-tp8-pp2-a97c9cb1
model      /containers/llm/models/GLM-5.3-Flash   (FP8, 305.79 GiB)
hardware   2 nodes x 8x NVIDIA A40 48GB (GA102, sm_86), ECC disabled
           (49140 MiB/GPU), PCIe, no NVLink
driver     610.43.02, CUDA UMD 13.3
cluster    Ray 2 nodes, --distributed-executor-backend ray
</code></pre>
<p>Relevant NCCL settings (GPU P2P is non-functional on this host: ACS redirection
on every PCIe bridge, and NCCL hangs in channel connection at P2P levels PHB
and PIX even after clearing ACS, so P2P is disabled):</p>
<pre><code>NCCL_ALGO=Ring
NCCL_PROTO=Simple
NCCL_P2P_DISABLE=1
</code></pre>
<h2>Serve command</h2>
<pre><code>vllm serve /containers/llm/models/GLM-5.3-Flash \
  --tensor-parallel-size 8 \
  --pipeline-parallel-size 2 \
  --distributed-executor-backend ray \
  --max-model-len 524288 \
  --gpu-memory-utilization 0.90 \
  --max-num-seqs 4 \
  --max-num-batched-tokens 8192 \
  --enable-prefix-caching \
  --enable-prompt-tokens-details \
  --disable-custom-all-reduce \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE","cudagraph_capture_sizes":[1,2,4,8,16],"max_cudagraph_capture_size":16}' \
  --speculative-config '{"method":"mtp","num_speculative_tokens":3}' \
  --enable-auto-tool-choice --tool-call-parser glm47 --reasoning-parser glm45 \
  --host 0.0.0.0 --port 8000 --served-model-name glm-5.3-flash-fp8
</code></pre>
<p>Engine reports:</p>
<pre><code>kv cache group sizes [640, 4, 640, 640, 640, 640]
kv lcm block sizes 640
Available KV cache memory: 19.62 GiB     (per GPU, x16)
</code></pre>
<p>Note the hybrid group structure: one group at block size 4 alongside five at
640. Two nodes are required because the FP8 checkpoint does not fit one
(single node reports <code>Available KV cache memory: -2.75 GiB</code> at 8x47.40 GiB),
so PP=2 is not optional here.</p>
<h2>Symptom</h2>
<p>Agent output (Pi harness) collapses mid-generation into:</p>
<pre><code>sorbackend/renovisorbackend/renovisorbackend/renovisorbackend/renovisorbackend/
renovisorbackend/renovisorbackend/renovisorbackend/renovisorbackend/renovisor...
</code></pre>
<p>repeating until the token budget is exhausted. HTTP 200, no server-side error.</p>
<h2>The speculative-decode metrics are the tell</h2>
<p>During normal generation, MTP acceptance tracks content difficulty — roughly
47–55% average during reasoning, rising to ~90% on fluent prose. During a
degeneration event, acceptance goes to near-perfect, because a constant-token
loop is trivially predictable:</p>
<pre><code>18:55:59  Per-position acceptance rate: 0.969, 0.941, 0.887   Avg 93.2%   (normal)
18:57:49  Per-position acceptance rate: 1.000, 1.000, 0.973   Avg 99.1%   (loop)
</code></pre>
<p><code>Prefix cache hit rate: 98.1%</code> throughout. No preemption or recompute events
appear in the log around the failure.</p>
<p>vllm#55506 reports the inverse (acceptance collapsing ~50% -&gt; ~26%) on
Qwen3.8-Flash-Next; the direction presumably depends on whether the poisoned
row is the one being measured, but the "acceptance stops tracking content"
signature is the same.</p>
<h2>Trigger conditions observed</h2>
<p>Matches vllm#55506's description (PP&gt;=2, <code>max_num_seqs&gt;=3</code>, threshold at 2-&gt;3
concurrent slots):</p>
<ul>
<li><strong>One agent: clean.</strong> Long sessions, no degeneration observed.</li>
<li><strong>Two agents: fails.</strong> Degeneration began shortly after a second client
connected. Notably the <em>already-running</em> request was the one that broke,
which is consistent with state being indexed by batch position — a
pre-existing request is corrupted when batch composition changes.</li>
</ul>
<h2>What was ruled out</h2>
<p>Each of these was tested separately with a synthetic harness (24 requests,
4 concurrent, staggered starts, strict JSON schema, long generations):</p>

change | result
-- | --
num_speculative_tokens: 3 | fails
num_speculative_tokens: 1 | fails identically (23/24)
prefix caching alone, no strict schema, 80k-token prefix, 4 concurrent | 20/20 clean

<p>The synthetic harness also surfaced a <strong>separate</strong> issue, reported separately:
XGrammar-constrained decoding conflicts with this model's reasoning delimiters
(<code>Failed to advance FSM</code> / <code>grammar rejected tokens [..., 154847, ...]</code>,
terminating requests with HTTP 500). Those token ids are GLM's own special
tokens. That is not this bug — it is loud, deterministic and 96% reproducible,
whereas the degeneration is silent and intermittent. Mentioned only so the two
are not conflated.</p>
<h2>Upstream fix</h2>
<p><a href="https://github.com/vllm-project/vllm/pull/55506">vllm-project/vllm#55506</a>:</p>
<blockquote>
<p><code>MambaSpecDecodeGPUContext</code> captures the block tables' raw data pointers
exactly once (<code>initialize_from_forward_context</code> is idempotent by design), and
the two align copy kernels resolved the table row [by batch row rather than
request slot].</p>
</blockquote>
<blockquote>
<p>The loop token is 1023, which is what the sampler deterministically emits for
an all-NaN logits row — the request's recurrent state has been poisoned and
never recovers. This is the same fault as the illegal-memory-access reports
in #54173: loud on sm_121, silent on sm_80 where the misdirected read stays
mapped.</p>
</blockquote>
<p>Possibly also relevant: this fork's #77 (mamba block size resolved too early,
triggered by prefix-cache reuse above ~3.6k cached tokens, silently seeds mamba
state from an arbitrary block and returns HTTP 200). Symptoms overlap; unclear
whether both apply here or only one.</p>
<h2>Ask</h2>
<p>Is vllm#55506 in scope for the backport, and does the GLM-5.3-Flash
(<code>Glm5NextTextLinearAttention</code>) path need the same request-slot indexing that
the Qwen3.8 path got?</p>
<p>Happy to test a branch — this cluster reproduces it readily, and sm_86 gives
the silent variant, which may be a useful complement to the sm_121 reports.</p>
<h2>Workarounds in use</h2>
<p>Both of these keep the FP8 checkpoint, which is the point of running two nodes:</p>
<ul>
<li>drop <code>--speculative-config</code> (PP alone is reported clean) — costs roughly
1.5x on decode throughput</li>
<li>drop <code>--enable-prefix-caching</code> (MTP alone is reported clean) — impractical
here, hit rate is 98% and prefill on Ampere is expensive</li>
</ul>
<p>Dropping PP is not an option: the FP8 checkpoint does not fit a single node.</p></body></html>
</body>
</html># [Bug]: GLM-5.3-Flash degenerates into a constant-token loop with PP=2 + MTP + prefix caching on sm_86 (upstream fix exists: vllm#55506)

## Summary

Serving `GLM-5.3-Flash` (FP8, `Glm5NextForConditionalGeneration`) across two
nodes at TP=8 / PP=2 with MTP-3 and prefix caching, requests intermittently
degenerate into an unbounded constant-token loop. Output is prompt-independent
once it starts and never recovers. The request returns HTTP 200; nothing is
logged.

This appears to be the same fault as
[[vllm-project/vllm#55506](https://github.com/vllm-project/vllm/pull/55506)](https://github.com/vllm-project/vllm/pull/55506)
("Index mamba spec-decode block tables by request slot, not batch row"), which
was reproduced on Qwen3.8-Flash-Next. That PR notes the failure is loud on
sm_121 as an illegal memory access but **silent on sm_80 where the misdirected
read stays mapped**. These are A40s (sm_86) and the behaviour is the silent
variant: corrupted output, no crash.

Filing here because the fork carries the affected hybrid/spec-decode paths and
has already taken related fixes (#63, #77, #49), and because GLM-5.3-Flash may
need the same treatment the Qwen path got.

## Environment

```
image      lazymio/vllm-backport:latest-sm86  (+ ray[default] layered on top;
                                               upstream image ships no ray)
fingerprint vllm-0.13.1-tp8-pp2-a97c9cb1
model      /containers/llm/models/GLM-5.3-Flash   (FP8, 305.79 GiB)
hardware   2 nodes x 8x NVIDIA A40 48GB (GA102, sm_86), ECC disabled
           (49140 MiB/GPU), PCIe, no NVLink
driver     610.43.02, CUDA UMD 13.3
cluster    Ray 2 nodes, --distributed-executor-backend ray
```

Relevant NCCL settings (GPU P2P is non-functional on this host: ACS redirection
on every PCIe bridge, and NCCL hangs in channel connection at P2P levels PHB
and PIX even after clearing ACS, so P2P is disabled):

```
NCCL_ALGO=Ring
NCCL_PROTO=Simple
NCCL_P2P_DISABLE=1
```

## Serve command

```
vllm serve /containers/llm/models/GLM-5.3-Flash \
  --tensor-parallel-size 8 \
  --pipeline-parallel-size 2 \
  --distributed-executor-backend ray \
  --max-model-len 524288 \
  --gpu-memory-utilization 0.90 \
  --max-num-seqs 4 \
  --max-num-batched-tokens 8192 \
  --enable-prefix-caching \
  --enable-prompt-tokens-details \
  --disable-custom-all-reduce \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE","cudagraph_capture_sizes":[1,2,4,8,16],"max_cudagraph_capture_size":16}' \
  --speculative-config '{"method":"mtp","num_speculative_tokens":3}' \
  --enable-auto-tool-choice --tool-call-parser glm47 --reasoning-parser glm45 \
  --host 0.0.0.0 --port 8000 --served-model-name glm-5.3-flash-fp8
```

Engine reports:

```
kv cache group sizes [640, 4, 640, 640, 640, 640]
kv lcm block sizes 640
Available KV cache memory: 19.62 GiB     (per GPU, x16)
```

Note the hybrid group structure: one group at block size 4 alongside five at
640. Two nodes are required because the FP8 checkpoint does not fit one
(single node reports `Available KV cache memory: -2.75 GiB` at 8x47.40 GiB),
so PP=2 is not optional here.

## Symptom

Agent output (Pi harness) collapses mid-generation into:

```
sorbackend/renovisorbackend/renovisorbackend/renovisorbackend/renovisorbackend/
renovisorbackend/renovisorbackend/renovisorbackend/renovisorbackend/renovisor...
```

repeating until the token budget is exhausted. HTTP 200, no server-side error.

## The speculative-decode metrics are the tell

During normal generation, MTP acceptance tracks content difficulty — roughly
47–55% average during reasoning, rising to ~90% on fluent prose. During a
degeneration event, acceptance goes to near-perfect, because a constant-token
loop is trivially predictable:

```
18:55:59  Per-position acceptance rate: 0.969, 0.941, 0.887   Avg 93.2%   (normal)
18:57:49  Per-position acceptance rate: 1.000, 1.000, 0.973   Avg 99.1%   (loop)
```

`Prefix cache hit rate: 98.1%` throughout. No preemption or recompute events
appear in the log around the failure.

vllm#55506 reports the inverse (acceptance collapsing ~50% -> ~26%) on
Qwen3.8-Flash-Next; the direction presumably depends on whether the poisoned
row is the one being measured, but the "acceptance stops tracking content"
signature is the same.

## Trigger conditions observed

Matches vllm#55506's description (PP>=2, `max_num_seqs>=3`, threshold at 2->3
concurrent slots):

- **One agent: clean.** Long sessions, no degeneration observed.
- **Two agents: fails.** Degeneration began shortly after a second client
  connected. Notably the *already-running* request was the one that broke,
  which is consistent with state being indexed by batch position — a
  pre-existing request is corrupted when batch composition changes.

## What was ruled out

Each of these was tested separately with a synthetic harness (24 requests,
4 concurrent, staggered starts, strict JSON schema, long generations):

| change | result |
|---|---|
| `num_speculative_tokens: 3` | fails |
| `num_speculative_tokens: 1` | fails identically (23/24) |
| prefix caching alone, no strict schema, 80k-token prefix, 4 concurrent | 20/20 clean |

The synthetic harness also surfaced a **separate** issue, reported separately:
XGrammar-constrained decoding conflicts with this model's reasoning delimiters
(`Failed to advance FSM` / `grammar rejected tokens [..., 154847, ...]`,
terminating requests with HTTP 500). Those token ids are GLM's own special
tokens. That is not this bug — it is loud, deterministic and 96% reproducible,
whereas the degeneration is silent and intermittent. Mentioned only so the two
are not conflated.

## Upstream fix

[[vllm-project/vllm#55506](https://github.com/vllm-project/vllm/pull/55506)](https://github.com/vllm-project/vllm/pull/55506):

> `MambaSpecDecodeGPUContext` captures the block tables' raw data pointers
> exactly once (`initialize_from_forward_context` is idempotent by design), and
> the two align copy kernels resolved the table row [by batch row rather than
> request slot].

> The loop token is 1023, which is what the sampler deterministically emits for
> an all-NaN logits row — the request's recurrent state has been poisoned and
> never recovers. This is the same fault as the illegal-memory-access reports
> in #54173: loud on sm_121, silent on sm_80 where the misdirected read stays
> mapped.

Possibly also relevant: this fork's #77 (mamba block size resolved too early,
triggered by prefix-cache reuse above ~3.6k cached tokens, silently seeds mamba
state from an arbitrary block and returns HTTP 200). Symptoms overlap; unclear
whether both apply here or only one.

## Ask

Is vllm#55506 in scope for the backport, and does the GLM-5.3-Flash
(`Glm5NextTextLinearAttention`) path need the same request-slot indexing that
the Qwen3.8 path got?

Happy to test a branch — this cluster reproduces it readily, and sm_86 gives
the silent variant, which may be a useful complement to the sm_121 reports.

## Workarounds in use

Both of these keep the FP8 checkpoint, which is the point of running two nodes:

- drop `--speculative-config` (PP alone is reported clean) — costs roughly
  1.5x on decode throughput
- drop `--enable-prefix-caching` (MTP alone is reported clean) — impractical
  here, hit rate is 98% and prefill on Ampere is expensive

Dropping PP is not an option: the FP8 checkpoint does not fit a single node.

**Comment by 714172724a-creator** [comment]
PP+MTP在官方vllm都还没实现吧。这个vllm-backport只是个人分支，更不可能实现。
PP and MTP have not been implemented in the official vllm yet. This vllm-backport is just a personal branch, and it is even more impossible to implement.
